Tensile/Components/ShiftVectorComponents.py

```diff
@@ -509,3 +510,225 @@ def __call__(self, writer, kernel, tP):
         writer.vgprPool.checkIn(mReg)
         writer.vgprPool.checkIn(tReg)
         return kStr
+
+
+class ShiftVectorComponentsMFMASourceSwap(ShiftVectorComponents):
+    kernel = {"EnableMatrixInstruction": True,
+              "SourceSwap": True}
+
+    """
+    Shift Vector Components d0,1
+    """
+    def __call__(self, writer, kernel, tP):
+        """ when we enable shift ptr with vectorwidth(2), we shift global read on edge block when size % vectorwidth != 0.
+            For example if M size == 3 vector width == 2, we want to do global read for [0-1] and [2-3].
+            But 3 is not in memory object, so we shift to do global read [0-1] and [1-2].
+            So buffer become [0, 1, 1, 2], assume result in register is same as input [0, 1, 1, 2]
+            We need to shift it back to [0, 1, 2].
+
+            In MFMA outputs, We have numContinuousOutput(4) for each thread.
+            We have numThreadInWave(64) threads.
+            number of thread in N is sames as kernel["MatrixInstN"] (32)
+            number of thread in M is numThreadInWave/numOutputThreads1 = 2
+            stride of continous output for each thread (numSubOutputPerWave0) is numOutputThreads0 * numContinuousOutput, (8).
+            we have numSubOutputGroupsPerWave0 which is 4 (kernel[tP["mt"]](64) // numSubOutputPerWave0(8))
+
+            So we do shift back by below alorithm.
+            1. check if M_size % GlobalLoadVectorWidth != 0, return if == 0
+            2. decide which subgroup we need to shift, M_size(3) means 3/8 = group 0
+            3. decide which thread we need to shift, we have different groups of thread, (0-31) for first group, (32-63) for second group.
+            4. decide which shift block (subTile1) we want to shift. for ex [0-1], [1-2], we want to shift second subtile
+        """
+
+        kStr = ""
+
+        regPerElem      = 2 if kernel["ProblemType"]["DataType"].isDouble() else 1
```

Comment:
kernel["ProblemType"]["ComputeDataType"].numRegisters() ?

Reply:
no,   d gemm equal to z, c gemm equal to s.